### ISSUE TITLE
SADX Graphics Options Modifications

### ISSUE THREAD

**Issue by ItsEasyActually** [X-Hax/SA-Mod-Manager#11]
~~Potential changes to make the user experience better. Notably changing how the full screen and windowed options are split along with the toggle for borderless full screen.~~

- [ ] ~~Change Windowed/Fullscreen options to a dropdown menu, including the Borderless option.~~

- [ ] ~~Brought up in #9, Fullscreen cannot go above the native resolution of the monitor normally. Locking the resolution options to max out at the selected window resolution is a solution to prevent this issue from occurring.~~

Update 11/25: Please see follow up comment for current progress, bugs, and plans.

**Comment by ItsEasyActually** [comment]
Active development can be followed in the dx-graphics-update branch. Overview followed by a checklist of changes and current problems below.

As it stands, this branch has changed the layout of the Graphics Config tab for SADX. It has also updated the GameSettings to v2 for SADX. This was due to a change in the overall handling of Profiles to be the point where both the SADXLoaderInfo and SADXConfigFile are now loaded and written from. Removes potential issues with having to bind to both SADXConfigFile and the Profile, and makes it so both the SADXLoaderInfo ini and SADXConfigFile can be written in a single function using the loaded Profile.

### Current Changes (Working):
- Graphics Config overhauled and simplified. 
  - Screen Mode (Formally the Windowed/Fullscreen radio buttons) has been added. It's a new dropdown, subject to change.
  - Simplified settings due to the aforementioned change. Only four graphics options exist now as Borderless and Custom Window were merged into the Screen Mode dropdown.
  - When Fullscreen is selected, the Render Resolution will be locked to the active screen's resolution. This will be updated to allow any resolution when Scale to Screen is also ticked as mentioned in the Planning section.
  - New strings were added or old ones updated to accommodate the new settings. Sorry translators :(
- SADX.GameSettings overhauled and updated to v2.
  - Added more variables to account for removing direct binding to SADXConfigFile.
  - Added dedicated Read/Write functions for SADXLoaderInfo and SADXConfigFile. 
  - Deprecated a handful of functions, referenced their new counterparts.

### Current Bugs:
- Borderless is currently non-functional. This will likely require a change to the Loader itself due to how the Borderless option is implemented. It also _seemingly_ does more than just make the window borderless which needs to be accounted for as well.

### Planned Additions/Modifications:
- [ ] Adjust so Render Resolution isn't locked if Scale to Screen is enabled. Arguably, remove the Scale to Screen option and have it be silently enabled if the Render Resolution is above the selected display's resolution.
- [ ] Update ToolTips for all updated Graphics elements and correct incorrect ones. 
- [ ] Fix Borderless issues. 
- [ ] Possibly Move Custom Window back to a CheckBox on the same row as Render Resolution.

**Comment by pkr-sadx** [comment]
Thanks for your hard work. I'll update the Mod Loader branch when the new configuration is finalized to make sure it reads all settings correctly. Here are some ideas and suggestions:

**Graphics tab**
1. The old "Borderless" checkbox makes the loader replace the WndProc to allow toggling between windowed mode and borderless fullscreen. This is a useful feature and should be preserved in some way. Maybe always replace WndProc in windowed mode? (though Resizable window has its own WndProc replacement).
2. "Pause when inactive" isn't a graphics option. I suggest moving it to Patches.
3. "Scale to screen" only works in when borderless fullscreen is enabled. Enabling "Scale to screen" makes it possible to do two things:
- Downsample from higher resolutions (on: fit the screen, off: display as-is so the content is cut off)
- Play at smaller resolutions (on: centered with a border, off: fill the whole screen)
I feel like this option should remain as a toggle, with an updated tooltip that explains it more clearly.

**Sound tab**
SFX volume control is only possible when BASS is enabled, so the slider should be grayed out if BASS for SFX is disabled.

**Comment by ItsEasyActually** [comment]
1. Borderless shouldn't be for anything related to Fullscreen, that option _should_ be for the disabling the window border and that's it. I think ours has been marred with that bad naming, but it always seems like it also just conflicts with what it should be doing as well. With it being a separate option users can select, we can have it and the custom window both replace the WndProc in whatever ways both need. I don't know if there'd be issues from always replacing it or not, but if there aren't, then we might as well. In fact, my earlier suggestion of moving the custom window back to a toggle would work two fold as I would think a borderless window could allow for resizing as well. 
2. I'm torn on moving it as it's a legacy fix that's existed in the Graphics tab for about as long as I can remember, and I also see it as a graphics fix. I'm not opposed to moving it; I'm just concerned for end users when this option is suddenly missing. The other issue with moving it to Patches means removing the option from the settings' Graphics class and moving it to a patch. In theory, should be fine, but I can see that being a bit of an issue. 
3. Scale to Screen should not require Borderless as those two options should be completely separate from each other. Scale to Screen should only matter if Fullscreen is enabled, and that's it. Since it does have an impact on Fullscreen resolutions as a whole, it will stay a toggle though. ToolTip needs to be updated, and if we can fix functionality with it, then it can be set to only be toggled when Fullscreen is the selected Screen Mode. Still debating on if we should use it as the means to unlock the Render Resolution for Fullscreen or let users potentially crash the game by having the render resolution above what their screen can normally handle. I'm personally in the camp of locking the res to the screen's max unless the option is ticked as I feel most programs would opt for something similar if they had such a feature. 

As for the sounds, I don't think I understand what you're getting at. In my quick test, there is a bug where the sliders for BGM and SFX aren't disabled when those options aren't selected. That's due to the settings change, so I do need to fix that. The BASS SFX slider is working as intended, and has no bearing on the other SFX slider.

**Comment by kellsnc** [comment]
Yeah 1 and 2 is the same behaviour as SA2 Mod Manager and I think it should be done that way.
For 3, I don't think we need scale to screen, downscaling/upscaling should be automatic.

**Comment by MainMemory** [comment]
As always, the current state of the mod loader is the result of different people adding features independently of one another.

**Comment by pkr-sadx** [comment]
Since the `json` branch of `sadx-mod-loader` was no longer mergeable without conflicts, I've picked all the changes I made manually and updated the master branch instead along with some other changes and fixes. The loader now reads the mods list and settings directly from the current profile's JSON file, and `SADXModLoader.ini` is no longer required.

**Comment by ItsEasyActually** [comment]
I've made a PR for the mod loader, referenced above. Once it's approved, we can wrap up the changes here and merge everything for the next release.

Main things of note is that Scale to Screen is just getting removed and defaulted to on when using anything that uses the parent/wrapper window. SA2's manager/loader already does this. The option is not really helpful for the end user and is just confusing in my opinion. 

I've also added an option to allow the mouse to be enabled when in fullscreen (either exclusive or borderless). It makes sense to me for this option to exist even if it doesn't have any major uses at the moment. The option is defaulted to off to stay in line with how things have been working.

I believe everything else is discussed in the PR above.
